compiler/msgs.nim

```diff
@@ -486,6 +486,39 @@ proc formatMsg*(conf: ConfigRef; info: TLineInfo, msg: TMsgKind, arg: string): s
               else: ErrorTitle
   conf.toFileLineCol(info) & " " & title & getMessageStr(msg, arg)
 
+proc colorError(s: string, color: ForegroundColor, doColor = false): string =
+  if doColor:
+    template isQuote(val: untyped): untyped = val == '\''
+    template isNotQuote(val: untyped): untyped = val != '\''
+    let parsable = (s.count('"').mod 2) == 0
+    var 
+      pos = 0
+    while pos < s.len:
+      if s[pos].isQuote:
+        if pos < s.high and s[pos + 1].isQuote:
+          result.add s[pos..(pos + 1)]
+        var inDoubleQuote = false
+        inc pos
+        let start = pos
+        while (pos < s.len and s[pos].isNotQuote) or inDoubleQuote:
+          if s[pos] == '"' and parsable:
+            inDoubleQuote = not inDoubleQuote
+          inc pos
+        #Highlight error
+        if pos < s.len and parsable:
+          if (s[pos].isQuote):
+            result.add fmt"""{color.ansiForegroundColorCode}{s[start..(pos - 1)]}{ansiResetCode}"""
```

Comment:
use sensible defaults; right now quoted expression looks too similar to Error when i tried it:
```nim
nim --colors -d:optUseColors --eval:asdf
```
(but see other comment: `-d:optUseColors` is a misuse, --colors should be all that's needed)


![image](https://user-images.githubusercontent.com/2194784/102337845-db55e700-3f47-11eb-9135-965e80a4e684.png)


=> D's choice of colors seems reasonable but anything distinctive is fine; we can distinguish:
current:
warning(yellow)
error(red)
hint(green)
new:
quoted expr(bold? cyan? magenta?)

Reply:
Well I see a `styleString` and a `quoteExpr` overload which takes the same params as `styleString` that means we can drop all `'` and then just call `quoteExpr` with our desired colour for each part, only thing is getting access to user defined colours and the conf as currently the procs that generate errors as far as i can see cannot access the `conf`.

Reply:
this works:
change to:
```nim
proc colorError(conf: ConfigRef, s: string, color: ForegroundColor): string =
  let doColor = optUseColors in conf.globalOptions
  let colorExpr = # get it from conf
```

`quoteExpr` doesn't need to have access to ConfigRef, all it does is quote expressions (and escaping quotes); btw: see how quoting works in addQuoted (which is for `"` instead of `'`)

Reply:
Well you've said `much as possible so code is correct "by construction"`, as far as I can see it's best to add styling when generating the messages, versus using escaped characters to get style after the message is created.

Reply:
> Well you've said

this is what I meant in https://github.com/nim-lang/Nim/pull/16356#discussion_r544537851 :
the following isn't correct by construction:
```nim
msg.add " foo '" # unmatched quote by itself
msg.add n.renderTree
msg.add "' rest of msg"
```
which is bad for 2 reasons:
* P1: msg is composed by half-matched atoms, and gets trickier with complex control flow (see sigmatch)
* P2: incorrect if renderTree contains some quote (would likely fail colorizing)

this is correct by construction:
```nim
msg.add " foo "
msg.add n.renderTree.quoteExpr
msg.add " rest of msg"
```
or even better:
```nim
msg.add " foo $1 rest of msg" % n.renderTree.quoteExpr
```

> as far as I can see it's best to add styling when generating the messages, versus using escaped characters to get style after the message is created
> best thing is to completely rewrite ...

The advantage of what I'm suggesting is that it requires minimum or no patching to most existing compiler code generating messages.

eg, this will automatically work with your PR (except for corner case where n.renderTree contains a quote):
```nim
msg.add " foo '$1' rest of msg" % n.renderTree
```
ideally this would be written as
```nim
msg.add " foo $1 rest of msg" % n.renderTree.quoteExpr
```
but it's not needed outside of above mentioned edge case.